chips/nrf5x/src/gpio.rs

```diff
@@ -504,26 +506,41 @@ impl<'a> hil::gpio::Interrupt<'a> for GPIOPin<'a> {
     }
 
     fn enable_interrupts(&self, mode: hil::gpio::InterruptEdge) {
-        if let Ok(channel) = self.allocate_channel() {
-            let polarity = match mode {
-                hil::gpio::InterruptEdge::EitherEdge => Config::POLARITY::Toggle,
-                hil::gpio::InterruptEdge::RisingEdge => Config::POLARITY::LoToHi,
-                hil::gpio::InterruptEdge::FallingEdge => Config::POLARITY::HiToLo,
-            };
-            let pin: u32 = (GPIO_PER_PORT as u32 * self.port as u32) + self.pin as u32;
-            self.gpiote_registers.config[channel]
-                .write(Config::MODE::Event + Config::PSEL.val(pin) + polarity);
-            self.gpiote_registers.intenset.set(1 << channel);
+        let channel = if let Some(chan) = self.allocated_channel.get() {
+            // We only support one interrupt mode per pin, despite the
+            // hardware supporting multiple. This is to comply with
+            // expectations in other Tock components, such as the button
+            // driver which re-registers interrupts for a restarted app,
+            // assuming the old ones to be overwritten.
```

Comment:
This change seems reasonable, but isn't the bigger issue that an app 1) made a request that allocated some resource and then 2) stopped (or restarted in this case) but didn't free the resource. So if an app stops you'll keep getting GPIO interrupts for it, even though there's no client?

I wonder if we handle this resource dropped issue _anywhere_ in Tock.

Reply:
Yes, this is an issue and I believe it to be an issue at large in the Tock ecosystem. We don't currently have a good way of testing or exercising these cases, and the fact that an app restart can trigger this relatively simple bug seems to be a symptom of that.

However, for this particular case, it's actually quite clear what we're ought to do:
https://github.com/tock/tock/blob/ae86c8366b064391571680c4f49508065a3d3569/capsules/core/src/button.rs#L245-L250

The driver itself is managing these subscriptions and multiplexing multiple applications onto interrupts. As such, when an application dies, there might actually be other apps that are still interested in this subscription. The driver does then exercise the above logic to deregister clients when they're no longer needed, lazily upon getting a callback.

The issue here is really just that we assume `enable_interrupt` to be an idempotent function, when it is not for the nRF. And that actually seems like a reasonable design to me, if only for this subsystem.